README.md
# NAME

MHA::AWS - A support script for "MySQL Master HA" running on AWS

# SYNOPSIS

    $ mhaws [subcommand] --interface_id=ENI-id [... args passed by MHA]
    $ mhaws [subcommand] --route_table_id=[RouteTable-id] --vip=[master VIP] [... args passed by MHA]

    required arguments:
      1. failover method is ENI attach/detach
        --interface_id=[ENI-id for master VIP]

      2. failover method is RouteTable change destination
        --route_table_id=[RouteTable-id]
        --vip=[master VIP]

    subcommand:
      master_ip_failover
      master_ip_online_change
      shutdown


    /etc/masterha_default.cnf

    [server default]
    master_ip_failover_script=mhaws master_ip_failover --interface_id=eni-xxxxxxxx
    master_ip_online_change_script=mhaws master_ip_online_change --interface_id=eni-xxxxxxxx
    shutdown_script=mhaws shutdown --interface_id=eni-xxxxxxxx

# DESCRIPTION

MHA::AWS is a support script for "MySQL Master HA" which running on Amazon Web Service.

# Installation

    # cpanm MHA::AWS

# REQUIREMENTS

- EC2 instance's "Name" tags must be resolved as DNS name in internal.
- root user must be allowed to ssh login between each MySQL instances.
- aws-cli is installed and available.

## Failover method = ENI attach/detach

- One ENI (Elastic Network Interface) must be attached to the MySQL master instance. Clients accesses for the ENI's IP address.

## Failover method = VPC route table rewriting

- Prepare a VIP address in your VPC.
- All MySQL hosts(master, slaves) can handle the VIP.
- Clients accesses for the VIP address.

# FAILOVER FLOW

1. MHA detect master failure.
2. "mhaws master\_ip\_failover --command stop",
    - (ENI) ENI will be detached from the old master instance.
    - (Route table) Route to VIP will be removed from VPC route table.
3. "mhaws shutdown --command (stopssh|stop)", Old master mysqld process will be killed (if ssh connection is available). Or old master instance will be stopped via AWS API (if ssh connection is NOT available).
4. MHA will elect the new master and set up replication.
5. "mhaws master\_ip\_failver --command start",
    - (ENI) ENI will be attached to the new master instance.
    - (Route table) Route to VIP will be set to new master instance.

# SEE ALSO

[AWS::CLIWrapper](https://metacpan.org/pod/AWS::CLIWrapper), [https://code.google.com/p/mysql-master-ha/](https://code.google.com/p/mysql-master-ha/)

# LICENSE

Copyright (C) FUJIWARA Shunichiro.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

FUJIWARA Shunichiro <fujiwara.shunichiro@gmail.com>


Changes
Revision history for Perl extension MHA-AWS

{{$NEXT}}

0.05 2014-08-16T08:09:50Z

   - update document for VPC route table rewriting.

0.04 2014-05-26T10:18:34Z

    - if timeout reached in stopping the old master instance (still "stopping" state), detected as it be stopped.

0.03 2014-05-19T11:01:50Z

    - fix status command

0.02 2014-05-19T08:27:36Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'MHA-AWS',
    module_name     => 'MHA::AWS',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/MHA/AWS/Shutdown.pm
package MHA::AWS::Shutdown;

use strict;
use warnings;
use parent "MHA::AWS";
use Log::Minimal;
use Net::SSH qw/ ssh /;

sub _command_stop {
    my $self = shift;

    my $res = $self->ec2("stop-instances", {
        instance_ids => $self->host_instance_id,
    });
    infof "result: %s", ddf $res;
    my $timeout = time + $MHA::AWS::API_APPLIED_TIMEOUT;
 STOP:
    while ( time < $timeout ) {
        $res = $self->ec2("describe-instances", {
            instance_ids => $self->host_instance_id,
        });
        my $instance = $res->{Reservations}->[0]->{Instances}->[0];
        if ( $instance->{State}->{Name} eq "stopped" ) {
            infof "instance stopped.";
            return 1;
        }
        sleep $MHA::AWS::CHECK_INTERVAL;
    }
    critf "TIMEOUT: %d sec. Can't confirm stop-instances: %s", $MHA::AWS::FAILOVER_TIMEOUT, $self->host_instance_id;
    return;
}

sub _command_stopssh {
    my $self = shift;

    my $target = sprintf "%s\@%s", $self->ssh_user, $self->host;
    my $command = q{
      killall -9 mysqld_safe mysqld
      for i in 1 2 3; do
        sleep 1
        pid=`pidof mysqld`
        if [ "x$pid" = "x" ]; then
          exit 0
        fi
      done
      exit 1
    };
    infof "ssh %s %s", $target, $command;
    my $r = ssh($target, $command);
    infof "exit code: %d", $r;
    return $r == 0 ? 1 : undef;
}

sub _command_status {
    my $self = shift;

    my $res = $self->ec2("describe-instances", {
        instance_ids => $self->host_instance_id,
    });
    my $instance = $res->{Reservations}->[0]->{Instances}->[0];
    if ( $instance->{State}->{Name} eq "running" ) {
        infof "instance is running.";
        return 1;
    }
    else {
        critf "instance state != running. %s", ddf $instance->{State};
    }
    return;
}

1;


lib/MHA/AWS/MasterIpFailover.pm
package MHA::AWS::MasterIpFailover;

use strict;
use warnings;
use parent "MHA::AWS";
use Log::Minimal;

sub _command_stopssh {
    my $self = shift;
    $self->_command_stop;
}

sub _command_stop {
    my $self = shift;
    if ($self->failover_method eq "eni") {
        $self->_command_stop_eni;
    }
    else {
        $self->_command_stop_route_table;
    }
}

sub _command_stop_eni {
    my $self = shift;

    my $res = $self->ec2("detach-network-interface", {
        attachment_id => $self->attachment_id,
    });
    infof "result: %s", ddf $res;
    my $timeout = time + $MHA::AWS::API_APPLIED_TIMEOUT;
  DETACHED:
    while ( time < $timeout ) {
        $res = $self->ec2("describe-network-interfaces", {
            network_interface_ids => $self->interface_id,
        });
        if ( $res->{NetworkInterfaces}->[0]->{Status} eq "available" ) {
            infof "detach completed.";
            return 1;
        }
        sleep $MHA::AWS::CHECK_INTERVAL;
    }
    critf "TIMEOUT: %d sec. Can't complete detach-network-interface: %s", $MHA::AWS::FAILOVER_TIMEOUT, $self->interface_id;
    return;
}

sub _command_stop_route_table {
    my $self = shift;

    my $destination_cidr_block = sprintf("%s/32", $self->vip);

    my $res = $self->ec2("delete-route", {
        route_table_id         => $self->route_table_id,
        destination_cidr_block => $destination_cidr_block,
    });
    infof "result: %s", ddf $res;
    my $timeout = time + $MHA::AWS::API_APPLIED_TIMEOUT;
 WAITING:
    while ( time < $timeout ) {
        $res = $self->ec2("describe-route-tables", {
            route_table_id => $self->route_table_id,
        });
        my @routes = @{ $res->{RouteTables}->[0]->{Routes} };
        for my $route (@routes) {
            if ( $route->{DestinationCidrBlock} eq $destination_cidr_block ) {
                sleep $MHA::AWS::CHECK_INTERVAL;
                next WAITING;
            }
        }
        infof "delete-route completed.";
        return 1;
    }
    critf "TIMEOUT: %d sec. Can't complete delete-route: %s %s", $MHA::AWS::FAILOVER_TIMEOUT, $self->route_table_id, $destination_cidr_block;
    return;
}

sub _command_start {
    my $self = shift;

    if ($self->failover_method eq "eni") {
        $self->_command_start_eni;
    }
    else {
        $self->_command_start_route_table;
    }
}

sub _command_start_route_table {
    my $self = shift;

    my $destination_cidr_block = sprintf("%s/32", $self->vip);

    my $res = $self->ec2("create-route", {
        route_table_id         => $self->route_table_id,
        instance_id            => $self->new_master_instance_id,
        destination_cidr_block => $destination_cidr_block,
    });
    infof "result: %s", ddf $res;
    my $timeout = time + $MHA::AWS::API_APPLIED_TIMEOUT;
 WAITING:
    while ( time < $timeout ) {
        $res = $self->ec2("describe-route-tables", {
            route_table_id => $self->route_table_id,
        });
        my @routes = @{ $res->{RouteTables}->[0]->{Routes} };
        for my $route (@routes) {
            if (
                $route->{DestinationCidrBlock} eq $destination_cidr_block
             && $route->{InstanceId} eq $self->new_master_instance_id
             ) {
                infof "create-route complated.";
                return 1;
            }
        }
        sleep $MHA::AWS::CHECK_INTERVAL;
    }
    critf "TIMEOUT %d sec. Can't complete create-route: %s %s", $MHA::AWS::API_APPLIED_TIMEOUT, $self->route_table_id, $destination_cidr_block;
    return;
}

sub _command_start_eni {
    my $self = shift;

    my $res = $self->ec2("attach-network-interface", {
        network_interface_id => $self->interface_id,
        instance_id          => $self->new_master_instance_id,
        device_index         => 2,
    });
    infof "result: %s", ddf $res;
    my $timeout = time + $MHA::AWS::API_APPLIED_TIMEOUT;
  ATTACHED:
    while ( time < $timeout ) {
        $res = $self->ec2("describe-network-interfaces", {
            network_interface_ids => $self->interface_id,
        });
        my $if = $res->{NetworkInterfaces}->[0];
        if ( $if->{Status} eq "in-use"
          && $if->{Attachment}
          && $if->{Attachment}->{Status} eq "attached"
          && $self->ping($self->vip)
        )
        {
            infof "attache completed.";
            return 1;
        }
        sleep $MHA::AWS::CHECK_INTERVAL;
    }
    critf "TIMEOUT %d sec. Can't complete attach-network-interface: %s", $MHA::AWS::API_APPLIED_TIMEOUT, $self->interface_id;
    return;
}

sub _command_status {
    my $self = shift;

    if ($self->orig_master_instance_id eq $self->current_attached_instance_id) {
        return 1;
    }
    else {
        critf "orig_master_instance_id: %s != current_attached_instance_id: %s", $self->orig_master_instance_id, $self->current_attached_instance_id;
        return;
   }
}

1;


lib/MHA/AWS.pm
package MHA::AWS;

use 5.008001;
use strict;
use warnings;
use Log::Minimal;
use AWS::CLIWrapper;
use Time::HiRes qw/ sleep /;
use Scalar::Util qw/ blessed /;
use Moo;

our $VERSION             = "0.05";
our $API_APPLIED_TIMEOUT = 120;
our $CHECK_INTERVAL      = 5;

has host             => ( is => "rw" );
has orig_master_host => ( is => "rw" );
has new_master_host  => ( is => "rw" );
has attachment_id    => ( is => "rw" );
has vip              => ( is => "rw" );
has ssh_user         => ( is => "rw" );
has instance_ids     => ( is => "rw", default => sub { +{} } );
has interface_id     => ( is => "rw" );
has failover_method  => ( is => "rw", required => 1 );
has route_table_id   => ( is => "rw" );
has aws => (
    is      => "rw",
    default => sub {
        AWS::CLIWrapper->new();
    },
);
has current_attached_instance_id => ( is => "rw" );

sub _init_eni {
    my $self = shift;
    my $res;
    $res = $self->ec2("describe-network-interfaces", {
        network_interface_ids => $self->interface_id,
    });
    my $interface = $res->{NetworkInterfaces}->[0];
    unless ($interface) {
        critf "Can't find network interface: %s", $self->interface_id;
        die;
    }

    $self->attachment_id( $interface->{Attachment}->{AttachmentId} );
    $self->current_attached_instance_id( $interface->{Attachment}->{InstanceId} );
    $self->vip( $interface->{PrivateIpAddress} );
}

sub _init_route_table {
    my $self = shift;
    my $res;
    my $destination_cidr_block = sprintf("%s/32", $self->vip);
    $res = $self->ec2("describe-route-tables", {
        route_table_id => $self->route_table_id,
    });
    if ( !$res->{RouteTables}->[0] ) {
        critf "Can't find route_table: %s", $self->route_table_id;
        die;
    }
    for my $route (@{ $res->{RouteTables}->[0]->{Routes} }) {
        if ($route->{DestinationCidrBlock} eq $destination_cidr_block) {
            $self->current_attached_instance_id($route->{InstanceId});
            return 1;
        }
    }
}

sub init {
    my $self = shift;

    if ($self->failover_method eq "eni") {
        $self->_init_eni;
    }
    else {
        $self->_init_route_table;
    }

    # create mapping table (hostname => instance-id) from tags
    my $res = $self->ec2("describe-tags");
    for my $tag (@{ $res->{Tags} }) {
        if ($tag->{ResourceType} eq "instance" && $tag->{Key} eq "Name") {
            $self->instance_ids->{ $tag->{Value} } = $tag->{ResourceId};
        }
    }

    if ( $self->new_master_host && !$self->new_master_instance_id ) {
        critf "Can't detect new_master_instance_id. abort";
        die;
    }
    if ( $self->orig_master_host && !$self->orig_master_instance_id ) {
        critf "Can't detect orig_master_instance_id. abort";
        die;
    }
}

sub ec2 {
    my $self = shift;
    my ($command, $args) = @_;
    infof "aws ec2 %s %s", $command, defined($args) ? ddf $args : "";
    my $res = $self->aws->ec2($command, $args);
    if ($res) {
        debugf "result: %s", ddf $res;
    }
    else {
        critf "failed";
        die;
    }
    $res;
}

sub ping {
    my $self    = shift;
    my $ip_addr = shift;
    system(qw/ timeout 3 ping -c 1 -t 3 /, $ip_addr) == 0 ? 1 : 0;
}

sub host_instance_id {
    my $self = shift;
    $self->instance_ids->{ $self->host };
}

sub new_master_instance_id {
    my $self = shift;
    $self->instance_ids->{ $self->new_master_host };
}

sub orig_master_instance_id {
    my $self = shift;
    $self->instance_ids->{ $self->orig_master_host };
}

sub dispatch {
    my $self    = shift;
    my $command = shift;
    if ( my $method = $self->can("_command_${command}") ) {
        infof "Invoke command: %s::%s", ref $self, $command;
        my $success = $method->($self);
        if ($success) {
            infof "Complete!";
        }
        else {
            infof "Failed!";
            die;
        }
    }
    else {
        critf("Unknown command: %s", $command);
        die;
    }
}

sub info {
    my $self = shift;
    my $info = {};
    for my $key (sort keys %$self) {
        next if Scalar::Util::blessed($self->{$key});
        $info->{$key} = $self->{$key};
    }
    $info;
}

1;
__END__

=encoding utf-8

=head1 NAME

MHA::AWS - A support script for "MySQL Master HA" running on AWS

=head1 SYNOPSIS

    $ mhaws [subcommand] --interface_id=ENI-id [... args passed by MHA]
    $ mhaws [subcommand] --route_table_id=[RouteTable-id] --vip=[master VIP] [... args passed by MHA]

    required arguments:
      1. failover method is ENI attach/detach
        --interface_id=[ENI-id for master VIP]

      2. failover method is RouteTable change destination
        --route_table_id=[RouteTable-id]
        --vip=[master VIP]

    subcommand:
      master_ip_failover
      master_ip_online_change
      shutdown


    /etc/masterha_default.cnf

    [server default]
    master_ip_failover_script=mhaws master_ip_failover --interface_id=eni-xxxxxxxx
    master_ip_online_change_script=mhaws master_ip_online_change --interface_id=eni-xxxxxxxx
    shutdown_script=mhaws shutdown --interface_id=eni-xxxxxxxx

=head1 DESCRIPTION

MHA::AWS is a support script for "MySQL Master HA" which running on Amazon Web Service.

=head1 REQUIREMENTS

=over 4

=item * EC2 instance's "Name" tags must be resolved as DNS name in internal.

=item * root user must be allowed to ssh login between each MySQL instances.

=item * aws-cli is installed and available.

=back

=head2 Failover method = ENI attach/detach

=over 4

=item * One ENI (Elastic Network Interface) must be attached to the MySQL master instance. Clients accesses for the ENI's IP address.

=back

=head2 Failover method = VPC route table rewriting

=over 4

=item * Prepare a VIP address in your VPC.

=item * All MySQL hosts(master, slaves) can handle the VIP.

=item * Clients accesses for the VIP address.

=back

=head1 FAILOVER FLOW

=over 4

=item 1 MHA detect master failure.

=item 2 "mhaws master_ip_failover --command stop",

=over 8

=item * (ENI) ENI will be detached from the old master instance.

=item * (Route table) Route to VIP will be removed from VPC route table.

=back

=item 3 "mhaws shutdown --command (stopssh|stop)", Old master mysqld process will be killed (if ssh connection is available). Or old master instance will be stopped via AWS API (if ssh connection is NOT available).

=item 4 MHA will elect the new master and set up replication.

=item 5 "mhaws master_ip_failver --command start",

=over 8

=item * (ENI) ENI will be attached to the new master instance.

=item * (Route table) Route to VIP will be set to new master instance.

=back

=back

=head1 SEE ALSO

L<AWS::CLIWrapper>, L<https://code.google.com/p/mysql-master-ha/>

=head1 LICENSE

Copyright (C) FUJIWARA Shunichiro.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

FUJIWARA Shunichiro E<lt>fujiwara.shunichiro@gmail.comE<gt>

=cut

